grass/kernel.c

```diff
@@ -61,16 +71,32 @@ void intr_entry(uint id) {
     if (earth->tty_recv_intr() && curr_pid >= GPID_USER_START) {
         /* User process killed by ctrl+c interrupt */
         INFO("process %d killed by interrupt", curr_pid);
-        asm("csrw mepc, %0" ::"r"(0x800500C));
+        proc_set[proc_curr_idx].mepc = (uint)sys_exit;
         return;
     }
 
     /* Ignore other interrupts for now */
     if (id == INTR_ID_SOFT) proc_syscall();
-    if (id == INTR_ID_TIMER) proc_yield();
+    proc_yield();
+}
+
+static void proc_request()
+{
+    int orig_idx = proc_curr_idx; // To Retain Scheduling Fairness
+
+    for (uint i = 0; i < MAX_NPROCESS; i++)
+        if (proc_set[i].status == PROC_REQUESTING){
+            proc_curr_idx = i;
+            earth->mmu_switch(curr_pid);
+            proc_syscall();
+        }
+
+    proc_curr_idx = orig_idx;
 }
 
 static void proc_yield() {
+    /* Run all requests to possibly make process runnable */
+    proc_request();
```

Comment:
I think we should move this to the for loop below. Specifically,
```c
if (s == PROC_READY || s == PROC_RUNNING || s == PROC_RUNNABLE) {
...
} else if (s == PROC_PENDING) { // Let's rename it to PROC_PENDING instead of PROC_REQUESTING
    earth->mmu_switch(proc_set[(proc_curr_idx + i) % MAX_NPROCESS].pid);
    proc_syscall();
}
```